PluginCore/PluginCore/Managers/EventManager.cs

```diff
@@ -95,26 +106,37 @@ public static void RemoveEventHandler(IEventHandler handler, EventType mask, Han
         public static void DispatchEvent(object sender, NotifyEvent e)
         {
             int length;
+            EventObject[] eventObjectsCopy;
             if (snapshotInvalid)
             {
-                length = highObjects.Count + normalObjects.Count + lowObjects.Count;
-                if (eventObjectsSnapshot.Length != length)
+                lock (eventLock)
                 {
-                    eventObjectsSnapshot = new EventObject[length];
+                    if (snapshotInvalid)
+                    {
+                        length = highObjects.Count + normalObjects.Count + lowObjects.Count;
+                        eventObjectsCopy = new EventObject[length];
+                        highObjects.CopyTo(eventObjectsCopy, 0);
+                        normalObjects.CopyTo(eventObjectsCopy, highObjects.Count);
+                        lowObjects.CopyTo(eventObjectsCopy, highObjects.Count + normalObjects.Count);
+                        eventObjectsSnapshot = eventObjectsCopy;
+                        snapshotInvalid = false;
+                    }
+                    else
+                    {
+                        eventObjectsCopy = eventObjectsSnapshot;
+                        length = eventObjectsCopy.Length;
+                    }
                 }
-                highObjects.CopyTo(eventObjectsSnapshot, 0);
-                normalObjects.CopyTo(eventObjectsSnapshot, highObjects.Count);
-                lowObjects.CopyTo(eventObjectsSnapshot, highObjects.Count + normalObjects.Count);
-                snapshotInvalid = false;
             }
             else
             {
-                length = eventObjectsSnapshot.Length;
+                eventObjectsCopy = eventObjectsSnapshot;
```

Comment:
Interlocked could be used... but the CLR spec enforces this to be atomic, so it shouldn't be  a problem...